flink-ml-lib/src/main/java/org/apache/flink/ml/feature/polynomialexpansion/PolynomialExpansion.java

```diff
@@ -0,0 +1,271 @@
+ * we get `(x, x * x, y, x * y, y * y)`.
+ *
+ * <p>For more information about the polynomial expansion, see
+ * http://en.wikipedia.org/wiki/Polynomial_expansion.
+ */
+public class PolynomialExpansion
+        implements Transformer<PolynomialExpansion>,
+                PolynomialExpansionParams<PolynomialExpansion> {
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    public PolynomialExpansion() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public Table[] transform(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        RowTypeInfo inputTypeInfo = TableUtils.getRowTypeInfo(inputs[0].getResolvedSchema());
+
+        RowTypeInfo outputTypeInfo =
+                new RowTypeInfo(
+                        ArrayUtils.addAll(inputTypeInfo.getFieldTypes(), VectorTypeInfo.INSTANCE),
+                        ArrayUtils.addAll(inputTypeInfo.getFieldNames(), getOutputCol()));
+
+        DataStream<Row> output =
+                tEnv.toDataStream(inputs[0])
+                        .map(
+                                new PolynomialExpansionFunction(getDegree(), getInputCol()),
+                                outputTypeInfo);
+
+        Table outputTable = tEnv.fromDataStream(output);
+        return new Table[] {outputTable};
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static PolynomialExpansion load(StreamTableEnvironment env, String path)
+            throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    /** Polynomial expansion function that expands a vector in polynomial space. */
+    private static class PolynomialExpansionFunction implements MapFunction<Row, Row> {
+        private final int degree;
+        private final String inputCol;
+
+        public PolynomialExpansionFunction(int degree, String inputCol) {
+            this.degree = degree;
+            this.inputCol = inputCol;
+        }
+
+        @Override
+        public Row map(Row row) throws Exception {
+            Vector vec = row.getFieldAs(inputCol);
+            if (vec == null) {
+                return row;
+            }
+            Vector outputVec = null;
+            if (vec instanceof DenseVector) {
+                int size = vec.size();
+                double[] retVals = new double[getResultVectorSize(size, degree) - 1];
+                expandDenseVector(((DenseVector) vec).values, size - 1, degree, 1.0, retVals, -1);
+                outputVec = new DenseVector(retVals);
+            } else if (vec instanceof SparseVector) {
```

Comment:
How about adding a `else` condition and throw exceptions when `!(vec instanceof DenseVector || vec instanceof SparseVector`?